dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/route/RouteService.java

```diff
@@ -224,24 +202,58 @@ public ResponseEntity<String> execute(
 
     String body = StreamUtils.copyToString(request.getInputStream(), StandardCharsets.UTF_8);
 
-    HttpEntity<String> entity = new HttpEntity<>(body, headers);
-
     HttpMethod httpMethod =
         Objects.requireNonNullElse(HttpMethod.valueOf(request.getMethod()), HttpMethod.GET);
     String targetUri = uriComponentsBuilder.toUriString();
 
+    WebClient.RequestHeadersSpec<?> requestHeadersSpec =
+        webClient
+            .method(httpMethod)
+            .uri(targetUri)
+            .httpRequest(
+                clientHttpRequest -> {
+                  Object nativeRequest = clientHttpRequest.getNativeRequest();
+                  if (nativeRequest instanceof HttpClientRequest httpClientRequest) {
+                    httpClientRequest.responseTimeout(
+                        Duration.ofSeconds(route.getResponseTimeoutSeconds()));
+                  }
+                })
+            .bodyValue(body);
+    for (Map.Entry<String, List<String>> header : headers.entrySet()) {
+      requestHeadersSpec =
+          requestHeadersSpec.header(header.getKey(), header.getValue().toArray(new String[0]));
+    }
+
     log.info(
         "Sending '{}' '{}' with route '{}' ('{}')",
         httpMethod,
         targetUri,
         route.getName(),
         route.getUid());
 
-    ResponseEntity<String> response =
-        restTemplate.exchange(targetUri, httpMethod, entity, String.class);
+    WebClient.ResponseSpec responseSpec =
+        requestHeadersSpec
+            .retrieve()
+            .onStatus(httpStatusCode -> true, clientResponse -> Mono.empty());
 
+    ResponseEntity<String> response =
+        responseSpec
+            .toEntity(String.class)
+            .onErrorReturn(
+                throwable -> throwable.getCause() instanceof ReadTimeoutException,
+                new ResponseEntity<>(HttpStatus.GATEWAY_TIMEOUT))
+            .onErrorResume(
+                throwable -> throwable.getCause() instanceof DataBufferLimitException,
+                throwable -> {
+                  JsonObject message =
+                      Json.object(
+                          jsonObjectBuilder ->
```

Comment:
I'd name this just `obj` or `object` because in the scope where the variable is valid it reads much better to pretend this is already an object. The "building" will happen implicitly at the end of the scope so it doesn't add anything to make the reader aware that the interface you are using is in fact a builder.